File: src/main/java/io/treutech/plainjson/StreamsProcessor.java
package io.treutech.plainjson;

import io.treutech.Constants;
import io.treutech.Person;
import org.apache.kafka.common.serialization.Serdes;
import org.apache.kafka.streams.KafkaStreams;
import org.apache.kafka.streams.KeyValue;
import org.apache.kafka.streams.StreamsBuilder;
import org.apache.kafka.streams.Topology;
import org.apache.kafka.streams.kstream.Consumed;
import org.apache.kafka.streams.kstream.KStream;
import org.apache.kafka.streams.kstream.KeyValueMapper;
import org.apache.kafka.streams.kstream.Produced;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.time.LocalDate;
import java.time.Period;
import java.time.ZoneId;
import java.util.Properties;

// $ kafka-topics --bootstrap-server localhost:9092 --create --topic ages --replication-factor 1 --partitions 4

public final class StreamsProcessor {
  private final Logger log = LogManager.getLogger(StreamsProcessor.class);

  private final String brokers;

  public StreamsProcessor(final String brokers) {
    super();
    this.brokers = brokers;
  }

  public static void main(final String[] args) {
    (new StreamsProcessor("localhost:9092")).process();
  }

  public void process() {
    final StreamsBuilder streamsBuilder = new StreamsBuilder();

    // Leemos desde el tópico de persons y deserializamos el valor como un objeto Person
    final KStream<String, String> personJsonStream = streamsBuilder.stream(
        Constants.getPersonsTopic(), Consumed.with(Serdes.String(), Serdes.String()));

    final KStream<String, Person> personStream = personJsonStream.mapValues((v -> {
      try {
        return Constants.getJsonMapper().readValue(v, Person.class);
      } catch (IOException e) {
        log.error(e.getMessage());
        return null;
      }
    }));

    // Mapeamos el flujo para calcular la edad y crear una nueva KeyValue con el nombre y la edad
    final KStream<String, String> ageStream =
        personStream.map((KeyValueMapper<String, Person, KeyValue<String, String>>) (k, v) -> {
          if (v == null) {
            return new KeyValue<>(k, null);
          }
          final LocalDate startDate = v.birthDate.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
          final int age = Period.between(startDate, LocalDate.now()).getYears();
          log.debug("Age: {}", age);
          return new KeyValue<>(v.firstName + " " + v.lastName, String.valueOf(age));
        });

    // Escribimos el flujo resultante en el tópico de ages
    ageStream.to(Constants.getAgesTopic(), Produced.with(Serdes.String(), Serdes.String()));

    final Topology topology = streamsBuilder.build();
    final Properties props = new Properties();
    props.put("bootstrap.servers", this.brokers);
    props.put("application.id", "kafka-examples");
    try (final KafkaStreams streams = new KafkaStreams(topology, props)) {
      streams.start();
    } catch (Exception e) {
      log.error("Error starting Kafka Streams: ", e);
    }
  }
}

File: src/main/kotlin/com/treutec/plain/StreamsProcessor.kt
package com.treutec.plain

import com.treutec.Person
import com.treutec.agesTopic
import com.treutec.jsonMapper
import com.treutec.personsTopic
import org.apache.kafka.common.serialization.Serdes
import org.apache.kafka.streams.KafkaStreams
import org.apache.kafka.streams.KeyValue
import org.apache.kafka.streams.StreamsBuilder
import org.apache.kafka.streams.kstream.Consumed
import org.apache.kafka.streams.kstream.KStream
import org.apache.kafka.streams.kstream.Produced
import org.apache.log4j.LogManager
import java.time.LocalDate
import java.time.Period
import java.time.ZoneId
import java.util.Properties

// kafka-topics --zookeeper localhost:2181 --create --topic ages --replication-factor 1 --partitions 4
// kafka-console-consumer --bootstrap-server localhost:9092 --topic ages --property print.key=true

fun main(args: Array<String>) {
    StreamsProcessor("localhost:9092").process()
}

class StreamsProcessor(val brokers: String) {

    private val logger = LogManager.getLogger(javaClass)

    fun process() {
        val streamsBuilder = StreamsBuilder()

        val personJsonStream: KStream<String, String> = streamsBuilder
                .stream(personsTopic, Consumed.with(Serdes.String(), Serdes.String()))

        val personStream: KStream<String, Person> = personJsonStream.mapValues { v ->
            val person = jsonMapper.readValue(v, Person::class.java)
            logger.debug("Person: $person")
            person
        }

        val resStream: KStream<String, String> = personStream.map { _, p ->
            val birthDateLocal = p.birthDate.toInstant().atZone(ZoneId.systemDefault()).toLocalDate()
            val age = Period.between(birthDateLocal, LocalDate.now()).years
            logger.debug("Age: $age")
            KeyValue("${p.firstName} ${p.lastName}", "$age")
        }

        resStream.to(agesTopic, Produced.with(Serdes.String(), Serdes.String()))

        val topology = streamsBuilder.build()

        val props = Properties()
        props["bootstrap.servers"] = brokers
        props["application.id"] = "kafka-examples"
        val streams = KafkaStreams(topology, props)
        streams.start()
    }
}